fs/inode/fs_files.c

```diff
@@ -583,6 +605,7 @@ int fs_getfilep(int fd, FAR struct file **filep)
   ret = nxmutex_lock(&list->fl_lock);
```

Comment:
can we replace fl_lock with critical section in all possible path?

Reply:
there have been objections from @pkarashchenko about this approach https://github.com/apache/nuttx/issues/6012#issuecomment-1597341056 and https://github.com/apache/nuttx/pull/10164#pullrequestreview-1574491947 .

But I'm not sure if these are correct. We have a potential thread suspend if `kmm_realloc` is called, but with `xxx_critical_section` I don't see the problem. The problem would be when using `spin_lock_xxx` which must never suspend, there is no such requirement for `xxx_critical_section`. 
Without `spin_lock_xxx` this problem will probably not be fixed for SMP, so it won't be the final solution.

Reply:
Another option is to use rwlock spinlock which is currently upstream for SMP case, and add a simple implementation of rwlock for non-SMP case where we use only critical sections. 

Can't rwlock for non-SMP be implemented simply as critical sections? Linux for rwlock in uniprocessor case simply disable preemption (

Reply:
can we call  kmm_xxx without critical section, and hold the critical section when updating fl_files since fl_files extension is an infrequent operation, we can free the alloced memory if other thread happen to extend fl_files successfully at the same time.

Reply:
It seems like there was no such problem before #3045, nuttx file descriptors predefined and static allocation at that time, there's no kmm_realloc and need no nxmutex_lock

Reply:
I still don't get why not call kmm_xx in critical section. I found 2 places in the kernel where we can find a similar practice
- https://github.com/apache/nuttx/blob/d1c73b6ed65cfb51dc8b8a8a8dc5854e77638f55/sched/task/task_setup.c#L143

- https://github.com/apache/nuttx/blob/d1c73b6ed65cfb51dc8b8a8a8dc5854e77638f55/sched/mqueue/msgget.c#L72

I guess these code work fine?

Reply:
kmm_xx may switch out due to fail to hold mm_lock, and then the critical section also auto leaves temporarily. Before the scheduler switch in this thread and rehold the critical section again, other threads may grip the critical section and enter the same code path.

Reply:
Thanks, I get it now.
But in that case I'll leave this PR as it is. It fixes a bug I'm familiar with and is's just simple.
We use fl_lock and critical section here but at least I know it works. Bigger change without serious testing have a high potential to break something and the possible bug can be difficult to detect.

Reply:
Ok, let's fix in the new patch.